text/0000-replace-slice.md

```diff
@@ -0,0 +1,180 @@
+- Feature Name: splice
+- Start Date: 2015-12-28
+- RFC PR:
+- Rust Issue:
+
+# Summary
+[summary]: #summary
+
+Add a `splice` method to `Vec<T>` and `String` removes a range of elements,
+and replaces it in place with a given sequence of values.
+The new sequence does not necessarily have the same length as the range it replaces.
+In the `Vec` case, this method returns an iterator of the elements being moved out, like `drain`.
+
+
+# Motivation
+[motivation]: #motivation
+
+An implementation of this operation is either slow or dangerous.
+
+The slow way uses `Vec::drain`, and then `Vec::insert` repeatedly.
+The latter part takes quadratic time:
+potentially many elements after the replaced range are moved by one offset
+potentially many times, once for each new element.
+
+The dangerous way, detailed below, takes linear time
+but involves unsafely moving generic values with `std::ptr::copy`.
+This is non-trivial `unsafe` code, where a bug could lead to double-dropping elements
+or exposing uninitialized elements.
+(Or for `String`, breaking the UTF-8 invariant.)
+It therefore benefits form having a shared, carefully-reviewed implementation
+rather than leaving it to every potential user to do it themselves.
+
+While it could be an external crate on crates.io,
+this operation is general-purpose enough that I think it belongs in the standard library,
+similar to `Vec::drain`.
+
+# Detailed design
+[design]: #detailed-design
+
+An example implementation is below.
+
+The proposal is to have inherent methods instead of extension traits.
+(Traits are used to make this testable outside of `std`
+and to make a point in Unresolved Questions below.)
+
+```rust
+#![feature(collections, collections_range, str_char)]
+
+extern crate collections;
+
+use collections::range::RangeArgument;
+use std::ptr;
+
+trait VecSplice<T> {
+    fn splice<R, I>(&mut self, range: R, iterable: I) -> Splice<I>
+    where R: RangeArgument<usize>, I: IntoIterator<Item=T>;
+}
+
+impl<T> VecSplice<T> for Vec<T> {
+    fn splice<R, I>(&mut self, range: R, iterable: I) -> Splice<I>
+    where R: RangeArgument<usize>, I: IntoIterator<Item=T>
+    {
+        unimplemented!() // FIXME: Fill in when exact semantics are decided.
+    }
+}
+
+struct Splice<I: IntoIterator> {
```

Comment:
I think this should be `Iterator` else it won't work out. But that's a technicality.